Added support for diff mode in tlconfig.

Thulium-Drake opened the pull request:
Hi there,

I added support for diff mode to the ```tlconfig``` module and while I was at it, I noticed that the output that came from the subprocess was in encoded-binary, which made it impossible to actually verify the current and desired situation.

I fixed that as well, which makes the module idempotent in our runs :-)

Thulium-Drake commented:
I have tested this succefully on a RHEL8 system that runs Python 3.6.8 and the options used in subprocess should still be supported by Python 3.11 according to the documenation :)

CendioOssman commented on library/tlconfig.py:
```diff
@@ -31,16 +32,18 @@
-    tlproc = subprocess.Popen([TLCONFIG, "%s" % param], stdout=subprocess.PIPE)
-    stdout, _ = tlproc.communicate()
-    return stdout.rstrip()
+    tlproc = subprocess.run([TLCONFIG, "%s" % param], stdout=subprocess.PIPE, universal_newlines=True)
```
`subprocess.run` requires Python 3.5, which is higher than the ThinLinc requirement on 3.4. So this would exclude some users. Might be best to hold off on this until the system requirements for ThinLinc have been increased.

Thulium-Drake replied:
I'll try to look at this a bit more with a colleague to see if I can make it work. Just to check, what is the timeline for Thinlincs support for Python 3.4?

CendioOssman replied:
Ideally, when SLES 12 is EOL next year. But it looks like we'll need to drop it a bit ahead of schedule at the end of this year to support the latest distributions.

Thulium-Drake replied:
Right, in that case I'll try to get a hold of my Python friend and see if we can make it work with the modules available in 3.4 as well

CendioOssman commented on library/tlconfig.py:
```diff
@@ -55,17 +58,24 @@ def main():
 
-    if getvalue(module.params['param']) != module.params['value']:
+    current=getvalue(module.params['param'])
+    new=module.params['value']
```
Spaces around `=` please, to stay consistent with the rest of the code.

Thulium-Drake replied:
Oops! Missed those :sweat_smile: 

Fixed!

CendioOssman commented on library/tlconfig.py:
```diff
@@ -55,17 +58,24 @@ def main():
+        result['changed'] = True
+
+        if module._diff:
+            result['diff'] = dict(
```
Why `dict()` and not the regular `{}` syntax?

Thulium-Drake replied:
No reason, I'm not a master snake charmer :), I was reading examples which mentioned ```dict()```. I have updated the notation :)

CendioOssman commented on library/tlconfig.py:
```diff
@@ -55,17 +58,24 @@ def main():
-    module.exit_json(changed=changed,
-                     param=module.params['param'],
-                     value=module.params['value'])
+    module.exit_json(**result)
```
This drops the `param` and `value` members from the result. Was that on purpose?

Thulium-Drake replied:
Yes, I didn't include results in the original post, but here's the output from diff:

```
TASK [thinlinc_server : Configure /vsmagent/agent_hostname] *********************************************************
--- before
+++ after
@@ -1,2 +1,2 @@
-thinlinc.zm1.element-networks.nl
+thinlinc.element-networks.nl
 ...

changed: [thinlinc.zm1.element-networks.nl] => changed=true 
```

The main issue I see is that it can be a bit confusing (the diff output isn't the prettiest)

```
TASK [thinlinc_server : Configure /vsmagent/agent_hostname] *********************************************************
--- before
+++ after
@@ -1,2 +1,2 @@
-thinlinc.zm1.element-networks.nl
+thinlinc.element-networks.nl
 ...

changed: [thinlinc.zm1.element-networks.nl] => changed=true 
  param: /vsmagent/agent_hostname
  value: thinlinc.element-networks.nl
```

I put them back in for now

CendioOssman commented on library/tlconfig.py:
```diff
@@ -55,17 +58,24 @@ def main():
-        changed = True
+        result['changed'] = True
+
+        if module._diff:
```
What about if check and diff mode are both active?

Thulium-Drake replied:
Good catch! I made a fix where if we're in check mode we just don't do anything in the ```setvalue``` function. I think that was the best way to handle it :)

CendioOssman reviewed (changes requested):
Thank you for your contribution! I think this looks like a nice addition.

I have some notes, though. Please have a look.

CendioOssman commented on library/tlconfig.py:
```diff
@@ -31,41 +32,52 @@
+        if module._diff:
+            result['diff'] = {
+                "before": yaml.safe_dump(current),
+                "after": yaml.safe_dump(new)
```
This looks very odd. Are you sure it's supposed to be a yaml structure inside a json structure? That seems very redundant.

Thulium-Drake replied:
I'm not too sure, I used an example module for making the diff output and it used ```yaml.safe_dump``` to format the output. However, I have taken it out and it still seems to work just fine.

(sorry, I'm not a :snake: wizard ;-) )

Thulium-Drake commented:
For completeness sake, this is what it will produce when running normally:
```
TASK [thinlinc_server : Configure /vsmagent/agent_hostname] **********************************************************
--- before
+++ after
@@ -1 +1 @@
-thinlinc.element-networks.nlkkk
\ No newline at end of file
+thinlinc.element-networks.nl
\ No newline at end of file

changed: [thinlinc.zm1.element-networks.nl]
```

And this is in verbose 1 (```-v```):

```
TASK [thinlinc_server : Configure /vsmagent/agent_hostname] **********************************************************
--- before
+++ after
@@ -1 +1 @@
-thinlinc.zm1.element-networks.nl
\ No newline at end of file
+thinlinc.element-networks.nl
\ No newline at end of file

changed: [thinlinc.zm1.element-networks.nl] => changed=true 
  param: /vsmagent/agent_hostname
  value: thinlinc.element-networks.nl
```

samhed commented:
What is the status here?

Thulium-Drake commented:
This couldn't be merged at the time as the change uses python facilities not present in python 3.4 and earlier and SLES12 was/is still supported.

But AFAIK, the work itself is done, so unless there's a need for additional changes, feel free to merge :-)

samhed commented:
According to their website, general support for SLES 12 ends 2024-10-31:
https://www.suse.com/lifecycle/#suse-linux-enterprise-server-12

Thulium-Drake commented:
What is the last version of Thinlinc to support this? That might be a good cutoff moment :-)

CendioOssman commented:
The next version of ThinLinc will be released after SLES 12 is EOL, so I don't think that is a blocker anymore. Our Python requirements will likely be 3.6 at that point.

Thulium-Drake committed:
Rebased

CendioOssman reviewed (approved):
Looks good! Thanks!

CendioOssman commented:
Merged as 9918f10695bbf19c847b1fdcd8840ee9cb024ab4 (adjusted the commit message back to the original one).
